### ISSUE TITLE
Port 86Box to Linux

### ISSUE THREAD

**Issue by joncampbell123** [86Box/86Box#136]
It's not clear what version to use. The configure script looks for allegro-config, which Allegro 5 (here on Github) doesn't seem to provide. Allegro 5 installs pkg-config files so you can use pkg-config instead.

I am unable to compile and develop this project until I can get this project's major dependency (Allegro) to compile and install for this project.

**Comment by OBattler** [comment]
Allegro is only needed for Linux, and right now the code won't compile for Linux anyway since too many changes were done to parts of the emulator while the Linux code has not yet been changed.
If you're compiling for Windows though, do NOT use configure, just run the MINGW32 environment and do make -fmakefile.mingw .

**Comment by joncampbell123** [comment]
Does that mean then that it would be trivial on my end to make a SDL-based frontend for Linux?

**Comment by MoochMcGee** [comment]
You'd need to figure out how to make config windows in SDL, but it is
possible, and would be appreciated.

On Sun, Jun 11, 2017 at 5:17 PM, Jonathan Campbell <notifications@github.com

**Comment by OBattler** [comment]
joncampbell123> You can port over the SDL/WxWidgets UI from mainline PCem. That should be easy.

**Comment by OBattler** [comment]
We (as in, the current 86Box team) had a discussion about this in the IRC channel last night, and we agreed we can try porting the SDL/WxWidgets UI from mainline PCem and adapting it to 86Box, but we need someone who understands that stuff to help.

**Comment by joncampbell123** [comment]
Certainly.

Decide which SDL library you want (2.x is recommended going forward, 1.x might be simpler).

SDL 2.x tutorial: http://www.willusher.io/pages/sdl2/

SDL 1.x tutorial: http://lazyfoo.net/SDL_tutorials/

**Comment by OBattler** [comment]
The UI in question uses SDL 2. I also got that UI's original author (bit) on board. See if you can reach him on the Vogons forum to coordinate the thing.

**Comment by mborjesson** [comment]
Hi joncampbell123,
I'm the author of the wxWidgets/SDL2-port used in PCem. Do you have any suggestions on how we should coordinate this work between us?

**Comment by joncampbell123** [comment]
@mborjesson Sure! Let me know what your git repository is, and I'll clone it and get to work. I'll submit patches or pull requests and we can work it out from there.

**Comment by mborjesson** [comment]
@joncampbell123 Great!
I haven't had time to really start with this project yet, but I've committed what I've done so far here:
https://github.com/mborjesson/86Box/tree/linux

**Comment by waltje** [comment]
Hey bit... would you mind contacting me (pref. on our IRC channel) regarding the Wx stuff?

**Comment by mborjesson** [comment]
Hey!

Of course! Unfortunately I won't have time today and probably not tomorrow but I should make it on Tuesday.
I'm sorry about the slow progress, I've had a lot less time available to spend on this than I initially thought I would have.

**Comment by waltje** [comment]
Hey bit, time for another conf call... :P

**Comment by waltje** [comment]
I am going to close this issue, since we are no longer going the Allegro route, but instead have a new Linux port available.

**Comment by OBattler** [comment]
I am reopening this because this needs to start again as waltje did not finish his own port and seems to have left the team without even giving us his unfinished wx code.

**Comment by DragonSWDev** [comment]
Is there any tl;dr information what would be needed to port it for Linux? I see it's already using SDL.

**Comment by dgurney** [comment]
> Is there any tl;dr information what would be needed to port it for Linux?

The biggest problem I can think of is the UI, which is currently not portable at all. There are plans to move to WxWidgets, which would solve that issue.

**Comment by ssokolow** [comment]
Has any consideration been put toward first releasing something more like QEMU and pretty much every old-school emulator build for Linux (eg. fceu, old-school snes9x, old-school MAME, etc.), where you have a GUI-less SDL build that's configurable via command-line arguments and manually editing the config file?

It wouldn't be ideal, but it'd at least be usable as a stop-gap solution. (Plus, it'd help for the automated testing uses I'm hoping to find a way to script.)

**Comment by dgurney** [comment]
It has been discussed in the past, but as far as I know that's as far as it got. In any case, contributions to help with Linux support in any capacity are certainly welcome, though it's probably wise to hop on Discord or IRC to discuss with others who are more involved than I am with the code.

**Comment by DragonSWDev** [comment]
@dgurney Yes, wxWidgets would bring real multiplatform GUI but I'm not sure how situation looks like with HiDPI. I remember that Dolphin developers switched from wxWidgets to Qt for some reasons and one of their reasons was better HiDPI support. Maybe Qt would be better idea? 

On the other hand PCEm is using wxWidgets so it would be probably easier to adopt it to 86box. I'm not sure what would be better option, as I said I'm not very experienced with Linux native GUI.

**Comment by SeanRamey** [comment]
Just a status update. I'm currently working on this.

My plan is to get it working natively on linux first, without a GUI, which means naking sure it can be fully run from the terminal. Then I'll work on the GUI for linux.So far, a GUI library hasn't been decided on. Looking at wxWidgets, Imgui, Nukclear.

This is going take a while because I don't have a lot of time to work on it, and I'm completely unfamiliar with 86box code. I'm doing it because nobody else is.

**Comment by waltje** [comment]
Already done.  It would be better to work on the UI part (I used Wx, but also have someone work on a Qt one)to merge them all.

    On Wednesday, May 5, 2021, 6:45:29 PM EDT, Sean Ramey ***@***.***> wrote:  
 
 


Just a status update. I'm currently working on this.

My plan is to get it working natively on linux first, without a GUI, which means naking sure it can be fully run from the terminal. Then I'll work on the GUI for linux.So far, a GUI library hasn't been decided on. Looking at wxWidgets, Imgui, Nukclear.

This is going take a while because I don't have a lot of time to work on it, and I'm completely unfamiliar with 86box code. I'm doing it because nobody else is.

**Comment by dhrdlicka** [comment]
It is quite hard not to duplicate efforts when all you've done with your Linux port is brag about it for several years straight without actually publishing it.

**Comment by joncampbell123** [comment]
Good to hear. I would do it but I'm too busy with DOSBox-X and my professional work and life. However once ported I might be able to submit patches for any remaining issues.

**Comment by qeeg** [comment]
@dhrdlicka LMAO REKT

**Comment by SeanRamey** [comment]
>Already done.  It would be better to work on the UI part (I used Wx, but also have someone work on a Qt one)to merge them all.

What? You've already done this? Why isn't it in 86box then? I don't understand.

**Comment by waltje** [comment]
Sean Ramey ***@***.***> wrote: 
> What? You've already done this? Why isn't it in 86box then? I don't understand.
Because I left the 86Box team before it was ready enough.
Its now part of my own project, and since I do not have time to 'manage' bothplatforms, I decided to focus on the Windows side first. I do keep it synced ofcourse, but no active dev on the Linux side for now.  Most of the dev work 
needed is Wx anyway, and not Linux per se.  It builds and runs just fine onMint etc.
--Fred

**Comment by SeanRamey** [comment]
> Because I left the 86Box team before it was ready enough. Its now part of my own project, and since I do not have time to 'manage' bothplatforms, I decided to focus on the Windows side first. I do keep it synced ofcourse, but no active dev on the Linux side for now.  Most of the dev work needed is Wx anyway, and not Linux per se.  It builds and runs just fine onMint etc. --Fred — You are receiving this because you modified the open/close state. Reply to this email directly, view it on GitHub, or unsubscribe.

Ok, so no plans to add it 86box anytime soon?

**Comment by waltje** [comment]
I wont (no time), but I am sure it would be ported to it.  Thats why I think it wouldbe a waste of time doing that again- instead, just focus on the UI part. I did a veryminimal Wx setup (so, just the basic menus and all that) so there's a fair bit of workleft there..
The good news: that UI work will also carry forward to Android and macOS, and,for a big part, even into a Qt port.
I wish they had (by now) created a UIdesigner application where you design the UI(in visual mode), and then click "Generate Code" where you select the platform forwhich you want it (Win32, GTK, Qt, Wx, etc etc) and voila..
--Fred

    On Wednesday, May 5, 2021, 10:24:27 PM EDT, Sean Ramey ***@***.***> wrote:  
 
 



Because I left the 86Box team before it was ready enough. Its now part of my own project, and since I do not have time to 'manage' bothplatforms, I decided to focus on the Windows side first. I do keep it synced ofcourse, but no active dev on the Linux side for now. Most of the dev work needed is Wx anyway, and not Linux per se. It builds and runs just fine onMint etc. --Fred — You are receiving this because you modified the open/close state. Reply to this email directly, view it on GitHub, or unsubscribe.


Ok, so no plans to add it 86box anytime soon?

**Comment by ssokolow** [comment]
> I wish they had (by now) created a UIdesigner application where you design the UI(in visual mode), and then click "Generate Code" where you select the platform forwhich you want it (Win32, GTK, Qt, Wx, etc etc) and voila..

There's too much variation in the available widget sets and how they're configured to make that an appealing thing for a hobby project.

Qt Designer's `.ui` files can be rendered to C++ using Qt's `uic` tool or to Python using PyQt's `pyuic` tool.

The GTK devs just deprecated Glade and now instruct people to write the GtkBuilder XML files by hand. I'm not aware of anything to render that to C at compile time, but I haven't actively looked for that.

wxWidgets has [various](https://wiki.wxwidgets.org/Tools) GUI builders but I don't know which (if any) support generating code rather than a runtime-loadable layout definition like Glade does for GTK.

I have no idea what's up with Win32 since my first GUI API after Visual Basic 6.0 was wxPython and, from there, I quit Windows for Linux in 2002 and then switched to PyGTK and then PyQt.

**Comment by waltje** [comment]
Well, it was just a hopeful thought ;-)
Yeah, I looked at some of the Wx builders, but most seem to want to use the'fluid' stuff.  Nice for phones and tablets, less so for regular machines.  The onethat is (or was) part of the Code::Blocks project was usable, that I remember. Ibelieve I even started implementing the UI in that, but then stopped dev for thereasons mentioned earlier.
So.. really.. we're talking about (G)UI development, mostly, and not the specificparts.
If someone wants to have a serious go at this, contact me.  As always, any ofthe VARCem code developed will be shared with 86Box, so there's no problemthere.
--Fred

    On Wednesday, May 5, 2021, 10:42:35 PM EDT, Stephan Sokolow ***@***.***> wrote:  
 
 



I wish they had (by now) created a UIdesigner application where you design the UI(in visual mode), and then click "Generate Code" where you select the platform forwhich you want it (Win32, GTK, Qt, Wx, etc etc) and voila..


There's too much variation in the available widget sets and how they're configured to make that an appealing thing for a hobby project.

Qt Designer's .ui files can be rendered to C++ using Qt's uic tool or to Python using PyQt's pyuic tool.

The GTK devs just deprecated Glade and now instruct people to write the GtkBuilder XML files by hand. I'm not aware of anything to render that to C at compile time, but I haven't actively looked for that.

wxWidgets has various GUI builders but I don't know which (if any) support generating code rather than a runtime-loadable layout definition like Glade does for GTK.

I have no idea what's up with Win32 since my first GUI API after Visual Basic 6.0 was wxPython and, from there, I quit Windows for Linux in 2002 and then switched to PyGTK and then PyQt.

**Comment by OBattler** [comment]
> I wont (no time), but I am sure it would be ported to it.
It wouldn't be, not with how many changes you're doing to the rest of the emulator, and the change already in the repository - Varcem's Win32 code is already incompatible with 86Box because of it. So it would take a LONG time to port, it's probably going to take less time to do it on our own.

**Comment by Cacodemon345** [comment]
So I took a shot at trying Linux support and while the UI parts have been stubbed and the platform-specific functions been dealt with, the only fish left to fry is the renderer. Unfortunately I have no idea how to make it work properly in such a multithreaded environment with SDL2. Right now I am thinking of just using the OpenGL code from the Windows backend with context sharing enabled to deal with threading issues since the SDL2 renderer path doesn't work well with multithreading on non-Windows and non-macOS platforms.

**Comment by OBattler** [comment]
I guess sticking to OpenGL only  for non-Windows platforms could work.

**Comment by Cacodemon345** [comment]
Actually I fixed the threading problem by queueing the blit request for the renderer thread to pick up.

**Comment by dob205** [comment]
I wonder how I can successfully compile the initial Linux/Unix code on macOS, I already have cmake and all dependencies ready.
For minivhd, slirp, and resid-fp it brings in a error message while running `cmake ../../src` after switching to the unix sub-directory inside the src directory:
```
CMake Error at disk/CMakeLists.txt:25 (target_link_libraries):
  Attempt to add link library "minivhd" to target "86Box" which is not built
  in this directory.

  This is allowed only when policy CMP0079 is set to NEW.


CMake Error at network/CMakeLists.txt:20 (target_link_libraries):
  Attempt to add link library "slirp" to target "86Box" which is not built in
  this directory.

  This is allowed only when policy CMP0079 is set to NEW.


CMake Error at sound/CMakeLists.txt:45 (target_link_libraries):
  Attempt to add link library "resid-fp" to target "86Box" which is not built
  in this directory.

  This is allowed only when policy CMP0079 is set to NEW.
```
Is there maybe something missing inside the cmake files or did I something wrong?

**Comment by Cacodemon345** [comment]
I don't have a Mac for testing macOS support.

**Comment by dob205** [comment]
> I don't have a Mac for testing macOS support.

On the other hand I'm here for testing macOS

**Comment by Cacodemon345** [comment]
Does it work now @dob205?

**Comment by dob205** [comment]
If I do `cmake ../../src` while being in the same directory it still struggles somehow, probably adding the same lines into the CMakeLists mentioned in the comment https://github.com/86Box/86Box/issues/136#issuecomment-903255127 and its error messages could work.

**Comment by 12101111** [comment]
> If I do `cmake ../../src` while being in the same directory it still struggles somehow, probably adding the same lines into the CMakeLists mentioned in the comment [#136 (comment)](https://github.com/86Box/86Box/issues/136#issuecomment-903255127) and its error messages could work.

```
cd 86Box && mkdir build && cd build && cmake -G Ninja .. -DCMAKE_BUILD_TYPE=Debug
```

![](https://user-images.githubusercontent.com/8438475/130356096-0686ea47-d7ea-4ba2-979b-188f2c74f0a2.png)

**Comment by dob205** [comment]
It does create now a working build chain, but on macOS it brings in following error:
```
/Users/myusername/86Box/src/config.c:2700:8: error: implicit declaration of function 'strnicmp' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                if (!strnicmp(hdd[c].fn, usr_path, strlen(usr_path)))
                     ^
/Users/myusername/86Box/src/config.c:2700:8: note: did you mean 'strncmp'?
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk/usr/include/string.h:84:6: note: 'strncmp' declared here
int      strncmp(const char *__s1, const char *__s2, size_t __n);
         ^

```

**Comment by Cacodemon345** [comment]
That one should be fixed too.

**Comment by dob205** [comment]
Mostly it works now, but stuff like memset (including <string.h> should fix that) and also having fseeko64, ftello64 and off64_t are breaking the compile at the moment. And also the compiler chokes on the function swab inside the slirp directory, otherwise it's working slowly towards a successful compile.

**Comment by Cacodemon345** [comment]
What are the errors?

**Comment by dob205** [comment]
The fseeko64, ftello64 and off64_t errors are mostly caused inside the hdd and the cdrom directory, because those function names seem to be outdated and should be replaced by fseeko, ftello and off_t.
As for the swab inside the slirp directory error following is happening right now on my Mac (seems to affect all C source files inside the directory, so only one example of the same error):
```
[1/349] Building C object src/network/slirp/CMakeFiles/slirp.dir/ip_icmp.c.o
FAILED: src/network/slirp/CMakeFiles/slirp.dir/ip_icmp.c.o 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -DCMAKE -DUSE_DYNAREC -DUSE_NEW_DYNAREC -I/Library/Frameworks/Mono.framework/Headers -I/usr/local/include/freetype2 -I/usr/local/Cellar/sdl2/2.0.16/include/SDL2 -I/Users/myusername/86Box/build/src/include -I/Users/myusername/86Box/src/include -I/Users/myusername/86Box/src/cpu -I/Users/myusername/86Box/src/codegen_new -O2 -g -DNDEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk -MD -MT src/network/slirp/CMakeFiles/slirp.dir/ip_icmp.c.o -MF src/network/slirp/CMakeFiles/slirp.dir/ip_icmp.c.o.d -o src/network/slirp/CMakeFiles/slirp.dir/ip_icmp.c.o -c /Users/myusername/86Box/src/network/slirp/ip_icmp.c
In file included from /Users/myusername/86Box/src/network/slirp/ip_icmp.c:34:
In file included from /Users/myusername/86Box/src/network/slirp/slirp.h:47:
In file included from /Users/myusername/86Box/src/network/slirp/util.h:34:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk/usr/include/unistd.h:597:10: error: conflicting types for 'swab'
void     swab(const void * __restrict, void * __restrict, ssize_t);
         ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk/usr/include/string.h:168:7: note: previous declaration is here
void     swab(const void * __restrict, void * __restrict, ssize_t);
         ^
1 error generated.
```

**Comment by qeeg** [comment]
That looks like a bug in macos's libc

**Comment by Cacodemon345** [comment]
That needs to be reported to Apple.

**Comment by dob205** [comment]
Do you mean the swab stuff @qeeg? Maybe there could be a workaround somehow, while it's reported towards Apple in parallel.

**Comment by dhrdlicka** [comment]
Thank you for the PR! I suppose the discussion directly related to it could be moved to that (#1626).

**Comment by dob205** [comment]
True, that makes sense @dhrdlicka.

**Comment by jvernet** [comment]
....swab... MacoSX...

It's not a MacosX Bug. It's because there is two or more #include that's colide. Try to remove string.h
Is the slirp code 64 bits clean  ? Otherwise it will crash badly on macosx.

**Comment by dob205** [comment]
I'll try that and also pull the fork from Cacodemon345 into my fork to get a easier way to do automatic compile testing.

**Comment by Cacodemon345** [comment]
> ....swab... MacoSX...
> 
> It's not a MacosX Bug. It's because there is two or more #include that's colide. Try to remove string.h

Are you sure that `<string.h>` include needs to be removed? Does any other header declare the functions in `<string.h>` on macOS?

**Comment by dhrdlicka** [comment]
What functions do you use from `string.h`?

**Comment by Cacodemon345** [comment]
I was referring to jvernet's comment.

**Comment by ata-star** [comment]
Thanks for the port, i like it
